crates/astria-core/src/protocol/transaction/v1/action/mod.rs

```
@@ -2579,8 +2580,8 @@ enum RecoverIbcClientErrorKind {
 
 #[derive(Debug, Clone)]
 pub enum CurrencyPairsChange {
-    Addition(Vec<CurrencyPair>),
-    Removal(Vec<CurrencyPair>),
+    Addition(BTreeSet<CurrencyPair>),
```

I wonder if this could be changed to use an `IndexSet` because these are usually faster than btrees. I think we are already having it as part of our public API anyway.

---

I went for the `BTreeSet` here since I thought having these ordered alphabetically would be more useful from a human-readable perspective.  `IndexSet` wouldn't do that (it retains the insertion order, more like a `Vec`) unless we explicitly call `sort_by` on it.

I doubt that performance is important here - we're likely dealing with small collections (so perf difference will be insignificant) and this action will likely be called very infrequently.

I could see a case to made for using `IndexSet` purely to retain the insertion order; it's maybe less surprising for a user creating an instance of this action.  But I tend to favour the alphabetical approach since it might be more obvious to the user that creating this action removes duplicates, which is important now that we fail execution for a duplicate entry.

---

Is there implications on the serialization order vs the execution here, and is there anything there that might matter?

IE I have a proto with an addition/removal action and a list of currency pairs. Those will be added or removed in alphabetical order instead of the order I specified them in?

---

I think the order in which these are _actually_ applied is invisible to the user.  The whole action will either succeed or fail as part of a tx: there's no exposure to a partial change whereby some of the additions/removals have been applied and some have not.

The only thing which I think might be surprising to the user is that (assuming the action is created via the proto object, e.g. via JSON rather than via e.g. a new CLI command) then the original order is changed to alphabetical and probably more importantly, duplicates have been removed.  But I'm not sure if there's anywhere the user has access to the parsed `Transaction` rather than the proto one.

---

Potentially it might be more honest to have the decoding fail if the proto contains duplicates, rather than just silently removing them for the user.  If we want to do that, then I guess `IndexSet` is back on the table, since there'd be nothing to warn the user about - we're no longer removing duplicates.

---

Changed to `IndexSet` and now fails to parse if the action contains duplicates.  Done in [a3ee841](https://github.com/astriaorg/astria/pull/2171/commits/a3ee841369ed28d328258c420a3066203a9c6bec).